src/util/newconfig/ConfigValue.hpp

```diff
@@ -83,47 +63,101 @@ class ConfigValue {
      * @return Reference to this ConfigValue
      */
     [[nodiscard]] ConfigValue&
-    defaultValue(Type value)
+    defaultValue(ValueType value)
     {
-        setValue(value);
+        auto const err = checkTypeConsistency(type_, value);
+        ASSERT(!err.has_value(), "{}", err->error);
+        value_ = value;
         return *this;
     }
 
     /**
-     * @brief Gets the config type
+     * @brief Sets the value current ConfigValue given by the User's defined value
      *
-     * @return The config type
+     * @param value The value to set
+     * @param key The Config key associated with the value. Optional to include; Used for debugging message to user.
+     * @return optional Error if user tries to set a value of wrong type or not within a constraint
      */
-    [[nodiscard]] constexpr ConfigType
-    type() const
+    [[nodiscard]] std::optional<Error>
+    setValue(ValueType value, std::optional<std::string_view> key = std::nullopt)
     {
-        return type_;
+        auto err = checkTypeConsistency(type_, value);
```

Comment:
Do you need to check type in constraints while there is the check here?

Reply:
I think so, because constraints applies to only a specific type, so if I do ConfigValue{Boolean}.withConstraint{port}
would be invalid

Reply:
Ah, after testing I remember the problem now. Imagine we have ConfigValue{Integer}.withConstraint{port} without a default value. Then the user must provide a value. Then, for example the user provides 123. We have to check to make sure 123 is of type Integer before then checking with constraint.